### DISCUSSION TITLE
Any way to integrate with Vercel AI SDK?

### DISCUSSION THREAD

**Discussion by Bartek532** [trpc/trpc#7271] [Q&A]
### Describe the feature you'd like to request

I'd like to integrate my tRPC serverless backend (made in Next.js) with Vercel AI SDK.

### Describe the solution you'd like to see

As I can see there is a pretty straightforward to integrate Next.js API Route handler with this SDK, so I wonder if we can just replace it with tRPC?

![image](https://github.com/user-attachments/assets/953d4e1e-1b32-48f1-ace9-3710c87b3bd0)

Seems like it only requires appropriate response to be returned, but is it possible to do it with tRPC, as it transforms all the outputs?

### Describe alternate solutions

We can use tRPC streaming together with the `ai` package, but then there is no possibility to use it together with hooks like `useChat` or `useCompletion`

### 👨‍👧‍👦 Contributing

- [ ] 🙋‍♂️ Yes, I'd be down to file a PR implementing this feature!

**Comment by BeBoRE** [comment]
You can do this with async generators https://trpc.io/docs/client/links/httpBatchStreamLink#generators

On the client-side this can be consumed with useQuery an array.

**Comment by Bartek532** [comment]
> You can do this with async generators https://trpc.io/docs/client/links/httpBatchStreamLink#generators
> 
> On the client-side this can be consumed with useQuery an array.

Yes, but can I plug it into Vercel AI SDK hooks somehow? Or I'm limited to Tanstack Query API?

**Comment by Floriferous** [comment]
Did someone get this to work correctly with streaming? Would be great to have a sample repo that shows this.

**Comment by sohanemon** [comment]
i got something working but it still doesn’t cover the full ai-sdk use case. here’s what i tried:

server (trpc):

```ts
openai: adminProcedure
  .input(
    z.object({
      text: z.string().min(1),
    }),
  )
  .mutation(async function* ({ input }) {
    const stream = streamText({
      model: ai.responses('gpt-4o-mini'),
      prompt: input.text,
    });

    yield* stream.textStream;
  })
```

client:

```tsx
const { mutateAsync, isPending } = api.admin.openai.useMutation({
  async onSuccess(data) {
    let fullResponse = '';
    for await (const val of data) {
      fullResponse += val;
    }
    setResponse(fullResponse);
  },
});
```

it streams fine through trpc, but it still doesn’t give the full ai-sdk features (like native client-side streaming helpers, aborting, ui bindings, etc). so it works, just not to the level the ai-sdk normally provides.

**Comment by danamajid** [comment]
Directionally this is what you'll want to do

1. Setup your trpc client to use batch stream link for your messages route, using split link
```ts
import {
  httpBatchStreamLink,
  httpBatchLink,
  splitLink,
} from "@trpc/client";
import { SuperJSON } from "superjson";

export function createTrpcClient() {
  return trpc.createClient({
    links: [
      splitLink({
        condition(opts) {
          return opts.path === "messages.send";
        },
        true: httpBatchStreamLink({
          url: apiUrl,
          transformer: SuperJSON,
        }),
        false: httpBatchLink({
          url: apiUrl,
          transformer: SuperJSON,
        }),
      }),
    ],
  });
}
```

2. Set up a custom chat transport
```ts
// Setup trpc chat transport
import { UIMessage, UIMessageChunk } from "ai";

type CustomChatConfig = {
  config: Record<string, unknown>;
  editingMessageId?: string;
};

// Transport wrapper for TRPC procedures
export function createTRPCChatTransport<
  TInput extends {
    chatId: string;
    message: UIMessage;
    messageId?: string;
    config?: CustomChatConfig;
    trigger: "submit-message" | "regenerate-message";
  },
  TOutput extends AsyncIterable<UIMessageChunk>,
>(
  trpcProcedure: (
    input: TInput,
    options?: { signal?: AbortSignal },
  ) => Promise<TOutput>,
) {
  return {
    async sendMessages(options: {
      trigger: "submit-message" | "regenerate-message";
      chatId: string;
      messageId: string | undefined;
      messages: UIMessage[];
      abortSignal: AbortSignal | undefined;
      body?: CustomChatConfig;
    }) {
      const message = options.messages[options.messages.length - 1];
      const result = await trpcProcedure(
        {
          chatId: options.chatId,
          message,
          messageId:
            options.trigger === "regenerate-message"
              ? options.messageId
              : options.body?.editingMessageId, // Support for editing messages
          trigger: options.trigger,
          config: options.body.config, 
        } as TInput,
        { signal: options.abortSignal },
      );

      // Convert AsyncIterable to ReadableStream
      const iterator = result[Symbol.asyncIterator]();
      const transformedStream = new ReadableStream({
        async pull(controller) {
          try {
            const { done, value: part } = await iterator.next();
            if (done) {
              controller.close();
              return;
            }

            controller.enqueue(part);
          } catch (error) {
            controller.error(error);
          }
        },
        async cancel() {
          await iterator.return?.();
        },
      });

      return transformedStream;
    },
    async reconnectToStream(_options: { chatId: string }) {
      // Return empty stream for reconnection - you can customize this if needed
      return new ReadableStream();
    },
  };
}
```

3. Define your route to yield and stream the response
```ts
import { UIMessage, createUIMessageStream } from 'ai';
import { z } from 'zod';
import { protectedProcedure, router } from '../trpc';

export const messagesRouter = router({
  send: protectedProcedure.input(z.object({
    chatId: z.string(),
    message: z.any() as z.ZodType<UIMessage>,
    trigger: z.enum(['submit-message', 'regenerate-message']),
    messageId: z.string().optional(), // For editing existing messages
    config: z.object(z.any()),
  }),
}).mutation(async function* ({ input }) {
  const stream = createUIMessageStream({
    execute: async ({ writer }) => {
      // Process tool mesages here
      writer.merge(input.message);
    },
  });

  return stream.toUIMessageStream();
});

```

4. Use custom transport with `useChat` in your chat component on the client
```ts
import { useChat } from "@ai-sdk/react";
const utils = trpc.useUtils();
const transport = createTRPCChatTransport(utils.client.messages.send.mutate);
const initialMessages = [];

const data = useChat({
  id: "123",
  messages: initialMessages,
  transport,
});
````

This should get you mostly there to support all of the ai sdk

**Comment by zirkelc** [comment]
Hi,

I created a demo repo for using AI SDK with TRPC:
- TRPC streaming mutations from an AI SDK UI message stream
- resume an an ongoing stream after reload/disconnect from a Redis in-memory server
- custom TRPC chat transport class for usage in `useChat`

Let me know if that helps you!

https://github.com/zirkelc/trcp-resumable-streaming/tree/main/src/examples/ai-sdk-trpc
